renderdoc/driver/ihv/intel/intel_counters.cpp

```diff
@@ -0,0 +1,423 @@
+  if(!maxFreqSymbol)
+    return 0;
+
+  TSetFrequencyOverrideParams_1_2 params;
+  params.Enable = true;
+  params.FrequencyMhz = maxFreqSymbol->ValueUInt32;
+  params.Pid = 0;
+  frequencyOverride->SetOverride(&params, sizeof(params));
+  Sleep(500);
+
+  return 0;
+}
+
+void IntelCounters::EndSesssion()
+{
+  if(!m_metricsDevice)
+    return;
+
+  TMetricsDeviceParams_1_2 *deviceParams = m_metricsDevice->GetParams();
+  if(deviceParams->Version.MajorNumber > 1 ||
+     (deviceParams->Version.MajorNumber == 1 && deviceParams->Version.MinorNumber >= 2))
+  {
+    IOverride_1_2 *frequencyOverride = m_metricsDevice->GetOverrideByName("FrequencyOverride");
+
+    if(frequencyOverride)
+    {
+      TSetFrequencyOverrideParams_1_2 params;
+      params.Enable = false;
+      params.Pid = 0;
+      frequencyOverride->SetOverride(&params, sizeof(params));
+    }
+  }
+}
+
+void IntelCounters::BeginPass()
+{
+  m_sampleIndex = 0;
+  TMetricSetParams_1_0 *metricSetParams = m_subscribedMetricSets[m_passIndex]->GetParams();
+  m_queryResult.resize(metricSetParams->MetricsCount + metricSetParams->InformationCount);
+}
+
+void IntelCounters::EndPass()
+{
+  m_passIndex++;
+}
+
+void IntelCounters::EnableCounter(GPUCounter index)
+{
+  uint32_t metricSetIdx = m_metricLocation[index].first;
+  IMetricSet_1_1 *metricSet = m_allMetricSets[metricSetIdx];
+  auto iter = std::find(m_subscribedMetricSets.begin(), m_subscribedMetricSets.end(), metricSet);
+  if(iter == m_subscribedMetricSets.end())
+  {
+    m_subscribedMetricSets.push_back(metricSet);
+    metricSetIdx = (uint32_t)m_subscribedMetricSets.size() - 1;
+  }
+  else
+  {
+    metricSetIdx = (uint32_t)(iter - m_subscribedMetricSets.begin());
+  }
+  m_subscribedMetricsByCounterSet[metricSetIdx].push_back(index);
+}
+
+void IntelCounters::DisableAllCounters()
+{
+  m_subscribedMetricSets.clear();
+  for(size_t i = 0; i < m_subscribedMetricsByCounterSet.size(); i++)
+    m_subscribedMetricsByCounterSet[i].clear();
+}
+
+void IntelCounters::BeginSample()
+{
+  if(!m_metricsDevice)
+    return;
+
+  D3D11_COUNTER_DESC counter_desc;
+  counter_desc.MiscFlags = 0;
+  counter_desc.Counter =
+      (D3D11_COUNTER)m_subscribedMetricSets[m_passIndex]->GetParams()->ApiSpecificId.D3D1XDevDependentId;
+
+  if(counter_desc.Counter == 0)
+    return;
+
+  TCompletionCode res = m_subscribedMetricSets[m_passIndex]->Activate();
+  if(res != TCompletionCode::CC_OK)
+    return;
+
+  HRESULT hr = m_device->CreateCounter(&counter_desc, &m_counter);
+  if(FAILED(hr))
+  {
+    m_subscribedMetricSets[m_passIndex]->Deactivate();
+    return;
+  }
+
+  res = m_subscribedMetricSets[m_passIndex]->Deactivate();
+  if(res != TCompletionCode::CC_OK)
+    return;
+
+  m_deviceContext->Begin(m_counter);
+}
+
+void IntelCounters::EndSample()
+{
+  if(!m_metricsDevice)
+    return;
+
+  m_deviceContext->End(m_counter);
+  HRESULT hr = S_OK;
+  uint32_t iteration = 0;
+  const uint32_t max_attempts = 0xFFFFFFFF;
```

Comment:
This value for `max_attempts` seems like it won't ever really fire if something does get stuck. Is there a better value we can pick? I don't know how many iterations on average are needed - since we sync the GPU I would guess a low number, so setting `max_attempts = 100` would perhaps be reasonable, but I don't have a sense of scale.